### ISSUE TITLE
GFX10 and GFX10.3 (Navi, RDNA) support in ACO

### ISSUE THREAD

**Issue by Venemo** [daniel-schuermann/mesa#136]
This issue is for tracking ACO's progress on Navi.

## What works, what doesn't

All shader stages should work. Every Vulkan game should work. 

If you find issues, please file a bug in the upstream Mesa bug tracker. 

## Tested hardware

- [x] Navi 10: Radeon RX 5700
- [x] Navi 10: Radeon RX 5700 XT
- [x] Navi 10: Radeon RX 5600 XT [tested and benchmarked by Phoronix](https://www.phoronix.com/scan.php?page=news_item&px=RX-5600-XT-New-vBIOS-Plus-ACO)
- [x] Navi 14: Radeon RX 5500 XT [tested and benchmarked by Phoronix](https://openbenchmarking.org/result/1912196-HU-EOY2019RA42&obr_sgm=y&obr_imw=y)
- [ ] Navi 12 should work, but not tested
- [x] Navi 2x: RX 6800 and 6800XT [tested and benchmarked by Phoronix] (https://www.phoronix.com/scan.php?page=article&item=rx6800-more-performance&num=1)

Not tested with unreleased Navi cards as we don't have those. If you test with hardware that is not on the list yet, please let us know.

## How to test

We suggest using the latest stable mesa, where ACO is the default compiler of the RADV Vulkan driver.

ACO is in mesa since version 19.3 but on old mesa releases, the `RADV_PERFTEST=aco` environment variable was needed.

## New hardware features support in Navi 1x

- [x] Wave32 (support for 32 lanes rather than 64)
- [x] NGG (Next Generation Geometry)

## New hardware features support in Navi 2x

- [ ] Hardware accelerated ray tracing
- [ ] Mesh shaders
- [ ] Variable rate shading

## Possible optimizations

- [ ] use round-robin register allocation to avoid WAR hazards (and help any post-RA scheduling)
- [ ] schedule ALU instructions (after RA for easier/faster scheduling?)
- [ ] choose registers to avoid bank conflicts (either as a reassignment pass or during RA)  
*See GCNRegBankReassign.cpp in LLVM*

- [ ] NGG shader based primitive culling

**Comment by SR-dude** [comment]
So, that's what the ACO developers have been doing for the past month.

**Comment by shmerl** [comment]
Just tested aco-navi branch with The Witcher 3 in Wine-esync+dxvk (Sapphire Pulse RX 5700 XT). It causes a GPU hang with this in dmesg (computer is accessible remotely through ssh):

```
[   52.097894] [drm:amdgpu_dm_atomic_commit_tail [amdgpu]] *ERROR* Waiting for fences timed out or interrupted!
[   57.207014] [drm:amdgpu_job_timedout [amdgpu]] *ERROR* ring gfx_0.0.0 timeout, signaled seq=2248, emitted seq=2251
[   57.207080] [drm:amdgpu_job_timedout [amdgpu]] *ERROR* Process information: process witcher3.exe pid 2612 thread witcher3.exe pid 2687
[   57.207083] [drm] GPU recovery disabled.
```

**Comment by Venemo** [comment]
@shmerl Basically most GPU hangs look like that in dmesg, so that message doesn't bring us closer to finding the problem. Can you try to identify which kind of shader is it that causes the hang? As a first step, can you try disabling CS in ACO? In `radv_pipeline.c` you can edit `radv_aco_supported_stage`, just comment out the CS support from there.

**Comment by Venemo** [comment]
Also please compile mesa in debug mode, just to see if it hits any assertions and such.

**Comment by Venemo** [comment]
Some fixes since yesterday:

* Fixed SLC, GLC, DLC bits in FLAT, Scratch and Global instructions
* Set DLC bit only for loads, don't set it according to GLC
* Added some RDNA ISA comments to ACO README
* Enabled SMEM SOE optimization on Navi (apparently it is supported even though there is no SOE bit)

**Comment by shmerl** [comment]
Compiled Mesa in debug mode without disabling CS first. It hangs too fast - I don't capture any output. Trying disabling CS now.

**Comment by shmerl** [comment]
Compiled with debug and captured some output. Without disabling CS:
```
info:  Game: witcher3.exe
info:  DXVK: v1.3.4
warn:  OpenVR: Failed to locate module
info:  Enabled instance extensions:
info:    VK_KHR_get_physical_device_properties2
info:    VK_KHR_surface
info:    VK_KHR_win32_surface
WARNING: Experimental compiler backend enabled. Here be dragons! Incorrect rendering, GPU hangs and/or resets are likely
WARNING: radv is not a conformant vulkan implementation, testing use only.
info:  AMD RADV/ACO NAVI10 (LLVM 10.0.0):
info:    Driver: 19.2.99
info:    Vulkan: 1.1.107
info:    Memory Heap[0]: 
info:      Size: 7920 MiB
info:      Flags: 0x1
info:      Memory Type[0]: Property Flags = 0x1
info:    Memory Heap[1]: 
info:      Size: 256 MiB
info:      Flags: 0x1
info:      Memory Type[2]: Property Flags = 0x7
info:    Memory Heap[2]: 
info:      Size: 8176 MiB
info:      Flags: 0x0
info:      Memory Type[1]: Property Flags = 0x6
info:      Memory Type[3]: Property Flags = 0xe
info:  D3D11CoreCreateDevice: Probing D3D_FEATURE_LEVEL_11_0
info:  D3D11CoreCreateDevice: Using feature level D3D_FEATURE_LEVEL_11_0
info:  Device properties:
info:    Device name:     : AMD RADV/ACO NAVI10 (LLVM 10.0.0)
info:    Driver version   : 19.2.99
info:  Enabled device extensions:
info:    VK_EXT_conditional_rendering
info:    VK_EXT_depth_clip_enable
info:    VK_EXT_host_query_reset
info:    VK_EXT_memory_priority
info:    VK_EXT_shader_demote_to_helper_invocation
info:    VK_EXT_shader_stencil_export
info:    VK_EXT_shader_viewport_index_layer
info:    VK_EXT_transform_feedback
info:    VK_EXT_vertex_attribute_divisor
info:    VK_KHR_create_renderpass2
info:    VK_KHR_dedicated_allocation
info:    VK_KHR_depth_stencil_resolve
info:    VK_KHR_descriptor_update_template
info:    VK_KHR_draw_indirect_count
info:    VK_KHR_driver_properties
info:    VK_KHR_get_memory_requirements2
info:    VK_KHR_image_format_list
info:    VK_KHR_maintenance1
info:    VK_KHR_maintenance2
info:    VK_KHR_sampler_mirror_clamp_to_edge
info:    VK_KHR_shader_draw_parameters
info:    VK_KHR_swapchain
info:  Device features:
info:    robustBufferAccess                     : 1
info:    fullDrawIndexUint32                    : 1
info:    imageCubeArray                         : 1
info:    independentBlend                       : 1
info:    geometryShader                         : 1
info:    tessellationShader                     : 1
info:    sampleRateShading                      : 1
info:    dualSrcBlend                           : 1
info:    logicOp                                : 1
info:    multiDrawIndirect                      : 1
info:    drawIndirectFirstInstance              : 1
info:    depthClamp                             : 1
info:    depthBiasClamp                         : 1
info:    fillModeNonSolid                       : 1
info:    depthBounds                            : 1
info:    multiViewport                          : 1
info:    samplerAnisotropy                      : 1
info:    textureCompressionBC                   : 1
info:    occlusionQueryPrecise                  : 1
info:    pipelineStatisticsQuery                : 1
info:    vertexPipelineStoresAndAtomics         : 0
info:    fragmentStoresAndAtomics               : 1
info:    shaderImageGatherExtended              : 1
info:    shaderStorageImageExtendedFormats      : 1
info:    shaderStorageImageReadWithoutFormat    : 0
info:    shaderStorageImageWriteWithoutFormat   : 1
info:    shaderClipDistance                     : 1
info:    shaderCullDistance                     : 1
info:    shaderFloat64                          : 1
info:    shaderInt64                            : 1
info:    variableMultisampleRate                : 1
info:  VK_EXT_conditional_rendering
info:    conditionalRendering                   : 1
info:  VK_EXT_depth_clip_enable
info:    depthClipEnable                        : 1
info:  VK_EXT_host_query_reset
info:    hostQueryReset                         : 1
info:  VK_EXT_memory_priority
info:    memoryPriority                         : 1
info:  VK_EXT_shader_demote_to_helper_invocation
info:    shaderDemoteToHelperInvocation         : 1
info:  VK_EXT_transform_feedback
info:    transformFeedback                      : 1
info:    geometryStreams                        : 1
info:  VK_EXT_vertex_attribute_divisor
info:    vertexAttributeInstanceRateDivisor     : 1
info:    vertexAttributeInstanceRateZeroDivisor : 1
info:  Queue families:
info:    Graphics : 0
info:    Transfer : 0
info:  DXVK: Read 471 valid state cache entries
info:  DXVK: Using 16 compiler threads
warn:  DXGI: VK_FORMAT_D24_UNORM_S8_UINT -> VK_FORMAT_D32_SFLOAT_S8_UINT
```
Hangs after that.

**Comment by shmerl** [comment]
When disabling CS, it hangs too fast, and capturing the output just produces an empty log for me.

Let me know, if you need the game to test it, may be developers can provide a key.

**Comment by aqxa1** [comment]
All DXVK/D9VK games that I've tested gpu hang on launch for me, so it might not be just a Witcher 3 issue.

Here's the list:
GTA4, GTA5, The Witcher 3, Mirror's Edge.

And, assuming I'm commenting out the correct line (stage == MESA_SHADER_COMPUTE), disabling CS has no effect.

I'm also using a custom card (MSI Evoke 5700xt) so maybe that's related.

**Comment by Venemo** [comment]
Thanks guys for your testing. I haven't tested any DXVK games yet so it's entirely possible that their shaders do something that aco-navi isn't prepared for. I'm currently working on implementing subgroup shuffles, but I promise I'll look into what is going on with DXVK.

**Comment by aqxa1** [comment]
@Venemo Actually, even vkcube causes a GPU hang for me.

With the following error:
amdgpu: radv_amdgpu_cs_query_fence_status failed.
amdgpu: The CS has been rejected, see dmesg for more information.
vk: error: failed to submit CS 0

Does that mean I failed to disable ACO's CS, or that Mesa's CS is hanging as well? It doesn't hang with normal LLVM Mesa, for the record.

**Comment by Venemo** [comment]
@aqxa1 @shmerl Are you guys sure that you disabled NGG during your testing? I think I mentioned in the original post that NGG is not implemented in ACO yet, but just to clarify I now edited the post and added a few env vars to "How to test".

This works for me without hanging:

```
RADV_DEBUG=nongg,nocache RADV_PERFTEST=aco vkcube
```

**Comment by aqxa1** [comment]
@Venemo That fixes the issue, thanks. I had assumed it just wasn't used rather than it needing to be explicitly disabled.

**Comment by aqxa1** [comment]
Just ran some quick tests. Other than some random GPU hangs, I also get the following error with TW3 and GTA5:
../src/amd/vulkan/radv_descriptor_set.c:496: VK_ERROR_OUT_OF_POOL_MEMORY

It's accompanied with missing/flickering models with GTA5 at least (I didn't test TW3 extensively).

**Comment by pendingchaos** [comment]
The VK_ERROR_OUT_OF_POOL_MEMORY error happens sometimes for me, I don't think it's an actual issue or related to ACO

**Comment by aqxa1** [comment]
> The VK_ERROR_OUT_OF_POOL_MEMORY error happens sometimes for me, I don't think it's an actual issue or related to ACO

I should add it doesn't occur with the same games under regular Mesa LLVM, and I get hundreds of them almost immediately when getting in game.

**Comment by pendingchaos** [comment]
Were you using a debug build? I think the error still happens with release builds but is only actually printed with debug builds.

If I'm remembering how DXVK handles descriptor pools correctly, this error is expected and DXVK handles it fine
EDIT: I might not be remembering correctly
EDIT2: I'm remembering correctly. DXVK allocates a new descriptor pool when the current one is out of memory

**Comment by Venemo** [comment]
@aqxa1 @shmerl If you guys still experience hangs or other problems with `nongg` then please give us a bit more details on what issues there are and how to reproduce those.

**Comment by aqxa1** [comment]
@pendingchaos You're right, it was just using a debug build that was caused the error messages.

@Venemo Would you prefer to add issues here, or open separate bugs?

**Comment by Venemo** [comment]
I think separate issues would be best. Just mention in those issues that they are about aco-navi.

**Comment by shmerl** [comment]
Thanks for the hint. Using `RADV_DEBUG=nongg,nocache` I was able to start TW3, so that hang was related. It still hangs sometime later, but at least it starts. I'll try to capture some output from the other hang later.

**Comment by shmerl** [comment]
Also, now that aco is merged into upstream, do we still need to use the external repo for testing (including for aco-navi?).

**Comment by pendingchaos** [comment]
This github repo will include some optimizations that haven't yet been upstreamed for a while
The contents of the aco-navi branch is not upstreamed, so you will need it to test ACO with Navi

**Comment by Venemo** [comment]
@shmerl I really appreciate your enthusiasm, but please keep in mind that aco-navi is still under heavy development, at this point I'm happy that it can run a bunch of example programs. (When I started, it could not even run the simplest triangle example.) This sort of work is not always as straightforward as it seems. I do plan to install the Witcher 3 on my computer and use that for testing, but only after fixing the issues that I currently know about. So no need to capture it.

**Comment by shmerl** [comment]
Great, thanks for the effort! I'll test it periodically then, to see how goes the progress.

**Comment by Venemo** [comment]
With regards to your question: at some point I'm gonna rebase my branch on upstream (after all the NIR stuff is merged, possibly), but I don't want to send aco-navi upstream until I'm satisfied that at least most of the popular games work well.

**Comment by wherron01** [comment]
Pardon me for being a bit clueless, but currently I’m running a 5700XT and am very interested in helping test aco-navi. i tried the mesa-aco-git package from lcarlier’s mesa-git repository, but obviously aco-navi isn’t on you guys’ master branch yet, and i got a GPU hang on boot. had to hard power cycle and boot straight into tty to fix, very messy. how would i go about installing aco-navi? is there an arch/aur/third party repo package i can install? i don’t really get how meson works and something about screwing with my graphics drivers seems a bit terrifying.

**Comment by Venemo** [comment]
@wherron01 The short answer is that it's not there yet. There is no easy way to install it unless you are comfortable compiling mesa on your own. Please give us a couple of weeks more to get it right. :)

**Comment by Venemo** [comment]
The aco-navi branch has gone through a major cleanup and is now rebased on top of latest mesa master (as of yesterday). There are still some issues, but overall it works much better for me. I haven't installed The Witcher 3 yet, but I did some testing with Dota 2.

**Comment by shmerl** [comment]
Thanks! I'll give it a go a bit later and will post results!

**Comment by shmerl** [comment]
Just tested The Witcher 3, it's still hanging after loading a saved game.

**Comment by aqxa1** [comment]
Updated tests with revision ~~51cab8b6990~~ e641024:

### Working

- Mirror's Edge (D9VK)
- Reshade (D9VK/DXVK) - effects ColorLookupTable, SMAA, CAS
- Hellblade : Senua's Sacrifice (DXVK)
- GTA4 (D9VK)
- GTA5 (DXVK)
- Persona 5 (RPCS3/Native) 
- Tokyo Xanadu Ex+ (DXVK)
- Grandia II Anniversary Edition (with HD remaster update) (DXVK)
- Crash Bandicoot N.Sane Trilogy (DXVK)
- Dying Light : Enhanced Edition (DXVK)
- Kingdom Come: Deliverance (DXVK)
- Sunset Overdrive (DXVK)
- Retroarch, Vulkan driver (Genesis Plus GX, BSNES)
- Prey (2017) (DXVK)
- The Witcher 3 (DXVK)
- Nier: Automata (DXVK)
- Hitman 2 (DXVK)
- The Witcher 2 (D9VK)
- Rise of the Tomb Raider (Native)
- Assassins Creed: Unity (DXVK)
- Thumper (DXVK)
- Deus Ex: Mankind Divided (DXVK)

### Hangs
- Retroarch (Mednafen PSX HW)
- ~~Dragon Quest XI (one hang fixed, one remaining) (DXVK)~~ (fixed by 4424784)
- ~~Assassin's Creed Unity (DXVK)~~ (fixed as of e641024)
- ~~Deus Ex : Mankind Divided (DXVK)~~ (fixed as of e641024)
- ~~Hitman 2 (DXVK)~~ (fixed by 4424784)
- ~~The Witcher 3 (DXVK)~~ (fixed by 4424784)
- ~~Yakuza : Kiwami (DXVK)~~ (fixed as of e641024)
- ~~Nier: Automata (DXVK)~~ (fixed by 4424784)
- ~~Thumper (DXVK)~~ (fixed as of e641024)

### Glitches
- Dragon Quest XI - draw-in/pop-in glitch (DXVK)
- ReShade - MXAO glitched (with GTA4)  (D9VK)

**Comment by pendingchaos** [comment]
>     * Mirror's Edge - this was working in older revisions, I just had forgot to set AMD_DEBUG=nongg
Did you mean RADV_DEBUG=nongg? Because AMD_DEBUG is for radeonsi, which ACO doesn't support

**Comment by aqxa1** [comment]
> > ```
> > * Mirror's Edge - this was working in older revisions, I just had forgot to set AMD_DEBUG=nongg
> > ```
> 
> Did you mean RADV_DEBUG=nongg? Because AMD_DEBUG is for radeonsi, which ACO doesn't support

Just a typo in the comment, which I have now fixed, thanks.

**Comment by Venemo** [comment]
Just a quick update on my progress here: I've spent the past few days on getting my patches in good shape so we can upstream them. While there are still issues, it can already run some games like Talos and Dota 2. But because there are issues, the plan is to keep ACO **disabled** on Navi upstream. I will continue working on it.

Currently I'm installing The Witcher 3.

**Comment by shmerl** [comment]
@Venemo: Did TW3 work for you or it's still hanging?

**Comment by Venemo** [comment]
@shmerl We've fixed a handful of issues, but TW3 still hangs. Trust me, you will hear about it when it works.

**Comment by Venemo** [comment]
Progress update: I'm now at the point where aco-navi can pass the CTS (Vulkan conformance test suite) almost the same as radv with llvm. There are currently two open merge requests: https://gitlab.freedesktop.org/mesa/mesa/merge_requests/2343 which fixes a couple of things that broke a lot of tests, and https://gitlab.freedesktop.org/mesa/mesa/merge_requests/2318 which adds subgroups support. I've also started looking into The Witcher 3 and found a problematic shader. Currently working on fixing that.

**Comment by Venemo** [comment]
@shmerl I believe I've got a fix for The Witcher 3, at least it doesn't hang for me anymore after loading the game. I'd appreciate if you could give a try to my latest aco-navi branch.

**Comment by shmerl** [comment]
Thanks! I'm going to test it shortly.

> Due to a bug in GFX10 hardware, s_nop instructions must be added
if a branch is at 0x3f. We already do this, but forgot to also update
the constant addresses that come after this instruction.

Do these workaround have any impact on performance?

**Comment by pendingchaos** [comment]
The 0x3f workaround shouldn't, it's needed infrequently and the s_nop should be very cheap

**Comment by shmerl** [comment]
I built it, but for some reason ACO is not showing up in the version anymore. Do I need to enable it with RADV_PERFTEST explicitly now?

UPDATE:

Yep, I had to use `RADV_PERFTEST=aco`.

Retesting now.

**Comment by shmerl** [comment]
Just tested it, and it's still causing a hang. From vulkaninfo for aco-navi build:

```
driverInfo = Mesa 19.3.0-devel (git-51cab8b699) (LLVM 10.0.0)
```
Save that's causing a hang.
[tw3_hang_save.zip](https://github.com/daniel-schuermann/mesa/files/3735912/tw3_hang_save.zip)

GPU: Sapphire Pulse RX 5700 XT.

**Comment by aqxa1** [comment]
I updated my comment with [new tests](https://github.com/daniel-schuermann/mesa/issues/136#issuecomment-538628059)  for the current aco-navi revision. Looking good with more games working properly, although I didn't test more than 15-30 minutes for each game.

**Comment by Venemo** [comment]
@aqxa1 can you please make a note in your comment about which of those are native and which are using DXVK or something else?

**Comment by Venemo** [comment]
@shmerl Thanks for trying. I had to upgrade my Witcher 3 but then I could load your savegame without issue. However, then it hanged within a minute as I was looking and walking around in the game. Is this what you've experienced?

**Comment by shmerl** [comment]
@Venemo : It hanged pretty much right away, doesn't even let walking around.

Game graphics settings are on max (hairworks off). Ambient occlusion: HBAO+.

**Comment by Venemo** [comment]
@shmerl But does the scene load up before it hangs? Because yesterday what I would see is that it hanged on the loading screen basically. Right now it loads the game and hangs some time after that. In my case, I'm running it on 3840x2160 but with the default settings.

EDIT: if I run it without debugging options, then it indeed hangs much sooner. if I run with `RADV_DEBUG=shaders,syncshaders` then it lasts a bit longer.

**Comment by shmerl** [comment]
It hanged right away after the loading screen, the scene didn't even show up. I built Mesa without debug though, so may be that's the reason.

Also, different settings can affect what shaders are used I suppose, so may be try max settings (hairworks off) / HBAO+.

**Comment by Venemo** [comment]
@shmerl I added a couple of patches that fix some so-called "hazards" on Navi. I haven't yet fixed every hazard that LLVM knows about, but still a good number of them. I compiled mesa in release mode, ran the Witcher 3 with `RADV_PERFTEST=aco`, set all graphics settings to ultra (except hairworks like you suggested), and then I set all postprocessing to maximum, and set HBAO+ too. Then I loaded your savegame and was able to walk around in the game for a couple of minutes without a hang. Can you confirm if these latest patches help you at all?

**Comment by shmerl** [comment]
Just built it and I was able to load the save without a hang! I didn't test it for long after that, but so far it looks good! Thanks for working on those.

One thing I noticed though, that aco performance looks just a tiny bit lower than llvm in that case:

**llvm**:

![tw3_llvm](https://user-images.githubusercontent.com/310927/67111083-32c0c180-f1a2-11e9-969f-5fa5c3a34937.jpg)

**aco**:

![tw3_aco](https://user-images.githubusercontent.com/310927/67111096-381e0c00-f1a2-11e9-8a37-d4e8ea4b0c06.jpg)

I hope it's not a consequence of those hazards mitigations.

Why are in general such "hazards" on Navi? Is it due to the need to support backwards compatibility with GCN ISA in RDNA?

**Comment by Venemo** [comment]
> One thing I noticed though, that aco performance looks just a tiny bit lower than llvm in that case

I think 110.2 vs 108.9 FPS is well within an acceptable margin, esp. considering that we haven't yet started working on any Navi-specific optimizations yet.

> Why are in general such "hazards" on Navi?

Every GPU has similar problems, but each generation has their own specific issues. Even though the hardware developers do their best to make sure the hardware works as intended, sometimes these sort of problems still slip in.

> Is it due to the need to support backwards compatibility with GCN ISA in RDNA?

The RDNA ISA is not backwards compatible with GCN. It is similar for sure, but you can't run the same binary on them.

**Comment by TacoDeBoss** [comment]
Is there a specified way to report test results? I've been playing Tetris Effect (non-steam) with ACO on my Navi 10 and it works beautifully.

**Comment by shmerl** [comment]
You can use the main benchmarks issue (#36), or may be there should be a separate one for Navi for clarity?

**Comment by Venemo** [comment]
@TacoDeBoss Thanks, I appreciate your feedback! I think we currently don't have a better way to report your results, but my collegaues will correct me there if I'm wrong.

**Comment by shmerl** [comment]
> The RDNA ISA is not backwards compatible with GCN. It is similar for sure, but you can't run the same binary on them.

Wasn't that the whole point why AMD still support Wave64 on Navi in less efficient way, while natively suggesting to use Wave32? I.e. I thought it's supposed to be able to run GCN targeted shaders fine, but also should work with ones that are RDNA only.  May be that kind of combination could potentially increase the probability of such hazards.

But anyway, I got your point about all GPUs potentially having such hardware issues.

**Comment by yshui** [comment]
I tried SteamVR home with aco-navi, which does not hang. But I noticed the whole scene looks much dimmer (lower brightness).

**Comment by Venemo** [comment]
@yshui Is this a navi specific problem?

**Comment by aqxa1** [comment]
Most GPU hangs look to be fixed by the hazards updates, at least the games I have tested. Dragon Quest XI looks to be completely hang free, at least in the places where LLVM still hangs.

**Comment by Venemo** [comment]
> Wasn't that the whole point why AMD still support Wave64 on Navi in less efficient way, while natively suggesting to use Wave32? I.e. I thought it's supposed to be able to run GCN targeted shaders fine, but also should work with ones that are RDNA only.

Well, it depends on what you mean by "compatible". What I meant was that a compiled shader that could run on Polaris or Vega can not run unmodified on Navi, so they are not compatible in this sense. This is not a suprising fact though, every GCN generation has (sometimes subtle) differences that makes it necessary to adapt the compiler to those generations.

If you are asking how close are the RDNA and GCN ISA, I think the difference between last gen GCN and RDNA is not that big. AMD could have called it GCN 6th gen, but maybe the marketing department decided they could use a refreshed branding. In case you are curious about the differences, you can take a look at the GFX10 commits in mesa.

> May be that kind of combination could potentially increase the probability of such hazards.

The hazards are usually a result of some rare corner case that the hardware designers forgot to take care about. Most of them fall into either of two categories:

1. The hardware implements an instruction in a buggy manner, so there are some use cases which don't produce the intended results
2. There are mistakes when it comes do dealing with dependencies between different hardware units that run in parallel, for example when one unit reads a register that is written to by the other, etc

> But anyway, I got your point about all GPUs potentially having such hardware issues.

Modern hardware is very complicated. It's not only GPUs that have problems, but everything else, eg. also CPUs. You can see a news item almost every week that talks about how some complicated sequence of instructions can cause unintended side effects; or how some instructions are unreliable on some processors.

**Comment by Venemo** [comment]
@aqxa1 Thanks for the feedback, it's nice to know that I'm heading in the right direction. Unfortunately those hazard mitigations are not perfect yet, and they don't yet take care of every possible occourence of the hazards. I plan to continue improving it next week.

**Comment by shmerl** [comment]
Further testing TW3, I found some anomaly with shadows:

**llvm:**

![shadows_llvm](https://user-images.githubusercontent.com/310927/67154560-5b44da80-f2cc-11e9-8907-fce0e27cfa6a.jpg)

**aco:**

![shadows_aco](https://user-images.githubusercontent.com/310927/67154565-60098e80-f2cc-11e9-80cf-048793532e7b.jpg)

Save at that place: 
[tw3_save_shadows_bug.zip](https://github.com/daniel-schuermann/mesa/files/3747451/tw3_save_shadows_bug.zip)

You can move forward using aco, and notice that shadows are missing in some distance.

**Comment by yshui** [comment]
@Venemo Sorry, but I am not sure. I don't have other hardware I can test on.

**Comment by Venemo** [comment]
@shmerl @yshui Since there have been other, similar reports from users on Polaris and Vega, I am inclined to think that these rendering glitches are not specific to aco-navi.

Since the discussion in this issue is starting to get very unwieldy, can I ask you guys to please file separate issues about these rendering glitches? (Make sure to still mention that you experienced the problem on aco-navi, it is useful to know.)

I would like to keep this issue about just "getting things to work" on Navi, in other words, ensure that most games can start and play without a hang or a crash. Once that's done, I will start looking at possible optimizations and individual rendering glitches.

**Comment by Venemo** [comment]
@aqxa1 Can you please take a look at the latest aco-navi and see if you still get any hangs?

**Comment by aqxa1** [comment]
@aqxa1 I updated my tests. I didn't test all of the games that were already fixed as of your previous hazard update, so I can't say if there are any regressions.

I will say that everything looks good, and Retroarch (Mednafen PSX HW) is the only application that seems to still be hanging. And it hangs with LLVM as well.

See [my previous comment](https://github.com/daniel-schuermann/mesa/issues/136#issuecomment-538628059).

**Comment by Venemo** [comment]
Thank you @aqxa1 I'm happy to hear that. The hazard mitigation series has now been reviewed by Daniel and Rhys, so I'm gonna land those upstream. With that taken care of, only the subgroups MR remains before we can enable the Navi support in ACO by default.

**Comment by Atrosha** [comment]
Can anyone post a guide or a link to guide of compile this branch for Steam on Ubuntu 19.04 x64? Steam needs 32-bit drivers/libs too.

**Comment by baryluk** [comment]
@Atrosha You can just wait a day or two, as it is going to be merged into mainline Mesa master - https://gitlab.freedesktop.org/mesa/mesa/merge_requests/2471

**Comment by baryluk** [comment]
@Atrosha You can take a look at my trivial script to build both 64-bit and 32-bit mesa with amdgpu support for Debian and Ubuntu. I am not sure it will work on Ubuntu, but it should be easy to modify.

https://gist.github.com/baryluk/1041204eff4cc4fad6f1508afe67b562

**Comment by shmerl** [comment]
Compiling 32-bit on 64-bit is a huge mess in Debian testing at least, unless you want to do it on 32-bit system.

For regular 64-bit build (on Debian, but could work for Ubuntu possibly), see: https://gist.github.com/shmerl/f4e5f76871239158cf083e37c5da56f4

To build Mesa-aco-navi, use it like this:

```mesa_branch='aco-navi' mesa_repo='https://gitlab.freedesktop.org/Venemo/mesa.git' src_dir="${HOME}/build/mesa-aco" mesa_dir='/opt/mesa-aco' mesa_debian_build.sh```

**Comment by shmerl** [comment]
@baryluk : I looked at your scripts. Interesting idea with using gcc-i686 toolchain. I can try that, since multilib method is totally borked.

**Comment by baryluk** [comment]
@shmerl It totally does work.

It should work on both Debian and Ubuntu. It took some time to figure out exactly which packages to install, and how to use proper `pkg-config` for 32-bit builds, env variables and cross files, but once you know it, it is straight forward. The only missing part for 32-bit build is libclc / libclang, but it is really needed just for OpenCL and even if enabled it will not work (there are other bugs in OpenCL implementation in Mesa). Filled a bug upstream - https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=941755

Other than that I use it daily to build mesa and 64-bit and 32-bit applications works just fine. I didn't want to make the script "too configurable", use Makefiles or functions (I don't see a point of having a function if it is only going to be called once), to make it straight forward to mess around, understand how it works or modify quickly.

I updated a script a little to be much easier to select LLVM and GCC versions, and made some package installation for amd64 and i386 less redundant and share common package selection automatically.

https://gist.github.com/baryluk/1041204eff4cc4fad6f1508afe67b562

Building aco-navi branch is one line change.

**Comment by shmerl** [comment]
I already filed a similar bug before: https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=890801

Anyway, apparently cross compiler has better situation than multilib one, so I'll try to update my script to use the former.

**Comment by Atrosha** [comment]
@baryluk at first I change Debian to LLVM repo in Ubuntu. Script uninstalls `glslang-dev` both: i386 and amd64 while compiling. And it needs to install gcc(g++)-9-i686 manually, coz installs version 8 by default.

So after successful compile `enable-new-mesa-opt.source` doesn't work on Ubuntu.

**Comment by baryluk** [comment]
@Atrosha I see. I updated the script a bit to install explicitly specific versions of gcc now.

glslang-dev is not required to build Mesa. I don't have it installed on my OS actually. But the script shouldn't make it remove it. In Debian and Ubuntu glslang-dev and glslang-tools package essentially have no dependencies so should not conflict with anything, even weird LLVM versions, and apt should never decide to remove it. So I don't know how it is possible it did remove glslang-dev.

The issue is probably somewhere else.

Could you download a new version from the gist and try again?

@shmerl  My script is still kind of not formally correct , because it actually uses 64-bit llvm-config for 32-bit build. But fortunately both 32-bit and 64-bit versions do output same linker and compiler flags, so it does work. If they would output different include paths, it would not work. But exactly same headers are used for both, so it is not a problem. AFAIK llvm-config doesn't understand multiarch, and it is not possible to install both 32-bit and 64-bit versions at the same time, and I wouldn't expect them to support that either.

**Comment by shmerl** [comment]
I just install one version at a time, and then remove it and install another one for different bitness build. But last time even doing that, co-installation of various stuff with impossible to get a consistent result using gcc-multilb. I'll try with cross compiler, to see how it fares.

You can specify 32-bit llvm in "native" file for meson.

**Comment by baryluk** [comment]
@shmerl  I never tried multilib. I just install cross-compilers, way easier and reliable, not to mention you can build also windows and arm binaries this way too. No you can't specificy 32-bit llvm-config in native file for meson native file because there is no 32-bit llvm-config available and co-installable with 64-bit llvm-config. But it is not a big deal. meson does have some disctinction between native and cross files, but I don't really use native. In fact I only use cross files even when compiling natively. But not to select architecture specific things, but to select llvm10 instead of other llvm versions possibly installed on the system.

**Comment by shmerl** [comment]
> No you can't specificy 32-bit llvm-config in native file for meson native file because there is no 32-bit llvm-config available and co-installable with 64-bit llvm-config

That's why I said, I remove one before installing the other. Not a big deal really. If co-installation of other stuff works, it should be OK.

**Comment by baryluk** [comment]
@shmerl Ah, yes. You can do that of course. I just find installing and removing one then another version constantly a bit annoying.

**Comment by shmerl** [comment]
Anyway, to avoid noise in this issue, I recommend anyone who needs help with building Mesa, to move the discussion [here](https://www.gamingonlinux.com/forum/topic/3396).

**Comment by baryluk** [comment]
@shmerl Agreed. Totally off topic.

**Comment by Venemo** [comment]
@shmerl @baryluk Sorry for my late reply, but if it's any help to you, here is a writeup about how I build mesa and run some apps using the built mesa: https://gist.github.com/Venemo/a9483106565df3a83fc67a411191edbd (without replacing the system package). This may need adjustments on Debian (such as /usr/lib64 and similar things), but should give you a good idea.

**Comment by shmerl** [comment]
@Venemo: Thanks, I'll take a look. The building part seems to be also using cross compiler for 32-bit, so I'll definitely go with that to finally have a script that builds both 64-bit and 32-bit. Using it without replacing the system packages looks quite similar to the approach I'm using: https://gist.github.com/shmerl/611b5b1670eb5963d0d99a4512ed8674

**Comment by baryluk** [comment]
@Venemo This is essentially what my scripts are doing, just fully automated plus both 64 and 32 bit build, opt and debug, plus not pollute system locations or even require root :D     https://gist.github.com/baryluk/1041204eff4cc4fad6f1508afe67b562    Thanks anyway.

**Comment by Zamundaaa** [comment]
As ACO for Navi seems to be enabled with the AUR aco repo now I tested on my Red Devil 5700 XT:
SteamVR
Blade & Sorcery
Gravity Lab
Pavlov VR
Witcher 3
I'll test some more in the next few days. When generating the shaders it was a little smoother (in VR that's quite noticable) and from there on not much difference, only a little stuttering in B&S was gone, which is nice. No graphics glitches or hangs.

**Comment by shmerl** [comment]
To clarify from where to get latest code for testing Navi with ACO. Should we still be using https://gitlab.freedesktop.org/Venemo/mesa/commits/aco-navi or more upstream https://github.com/daniel-schuermann/mesa/commits/master ?

**Comment by Venemo** [comment]
I recommend to just use [upstream mesa](https://gitlab.freedesktop.org/mesa/mesa) which has everything that was in aco-navi, and almost everything that's in Daniel's repo. Depending on your preference with regards to stability, you can go with either 19.3 RC, or master.

EDIT: to reduce the confusion, I now deleted the old aco-navi branch.

**Comment by shmerl** [comment]
What about ACO bug tracking, progress tracking and etc. Will that remain on Github, or also will be moved to Mesa Giltab?

**Comment by Venemo** [comment]
I will respond to any bug you might find in ACO on Navi, regardless of which place you report it. If there is a bug report that I don't notice feel free to ping me.

Not sure how long Daniel wants to maintain this repo, I assume on the long run all of it will move to upstream Mesa. @daniel-schuermann what is your stance on this? Personally, I have a slight preference towards using Mesa Gitlab, because it makes it easier to track which commits fix which problem.

**Comment by daniel-schuermann** [comment]
As some packages are based off this repo, we will maintain it for a little longer. It also contains one last performance-improving patch series from @pendingchaos which is not upstream yet, but should land soon. That said, I think with mesa 20.0 we'll fade it out into stable.
Bug tracking and progress tracking will probably stay even further than that, but hopefully, there won't be much left at that point.

**Comment by baryluk** [comment]
@daniel-schuermann Maybe announce a closure of accepting NEW issues on this repo (~Jan 2020?), and only deal with existing ones until they are closed. For all the other stuff redirect people to upstream mesa.

**Comment by Venemo** [comment]
We've recently landed Wave32 support in ACO. Those willing to test it can do so using `RADV_PERFTEST`. AMD recommends using this mode for compute shaders and the geometry engine. In case of pixel shaders, the old Wave64 mode is recommended since those shaders are more efficient in this mode due to how interpolation and the cache works in the hardware.

In my testing, `cswave32` and `gewave32` yield a couple of extra FPS in some games. This is not yet fully optimized, [Rhys has an open MR](https://gitlab.freedesktop.org/mesa/mesa/merge_requests/3019) which should give it a nice boost, once merged. The `pswave32` on the other hand reduces overall performance and is not recommended.

If you want to test, you need to use upstream mesa with `RADV_PERFTEST`

```
# The recommended way is to use Wave32 mode for compute shaders and the geometry engine
RADV_PERFTEST=aco,cswave32,gewave32 ./Talos
```

**Comment by Venemo** [comment]
In the meantime I'm working on a post-RA scheduler (based on Daniel's old code) with the intention of taking advantage of Navi's ILP (instruction level parallelism). The idea is to reorder the instructions (if possible) in such a manner that allows to execute as many as possible in parallel.

There are some issues that I'd like to resolve before opening a MR, such as compilation speed, it currently slows down ACO by about 10% which is not acceptable. However I can already confirm that it has a positive effect, I managed to squeeze out some extra frames per second from a couple of games. Will update you guys with some more concrete benchmarks once the code gets closer to final.

**Comment by Venemo** [comment]
This issue has been quiet for a while, so here is a quick heads up:

* Rhys has implemented geometry shaders support, which also works on Navi.
* He also made a couple of optimizations and took care of VOP3 literals and such.
* I'm now working on getting NGG working. So far I have some code which makes NGG VS work, next up on my TODO list is NGG GS, among other things.

**Comment by shmerl** [comment]
@Venemo: Do you see any performance improvements from NGG in your tests?

**Comment by Venemo** [comment]
@shmerl It's hard to say, because so far I've only got NGG VS covered. The "big fish" will be NGG GS. It will definitely provide some improvement, but I don't yet know how much.

However, considering your benchmark results so far, I wouldn't hold my breath that this will fix The Witcher 3 for you. One of us will have to take a close look at that game to determine where the bottleneck is. (I suspect that the problem exists in DXVK, because DXVK on Windows already performs 20% worse than D3D11, but I don't know what the problem is for sure.)

**Comment by shmerl** [comment]
@Venemo: It did get slightly better than llvm on Navi after recent batch of optimizations, but I suspect it's using a lot of tessellation shaders, so it will benefit more once those are implemented in ACO.

**Comment by Venemo** [comment]
@shmerl I ran some benchmarks yesterday with ACO NGG VS together with LLVM NGG VS+GS (as ACO doesn't support the merged VS+GS yet). It did not give me any noticable difference for any game that I tested with, compared to the legacy (non-NGG) pipeline. EDIT: I tested with Talos, RotTR and TW3.

Regardless of the performance implications, we'll probably eventually still need to support NGG, since it's not clear what will become of the legacy path in future chips. However it appears that NGG is **not** a silver bullet that would magically elevate our frame rates.

I suspect that the bottleneck lies with some very heavy "monster" pixel shaders rather than the geometry engine.

**Comment by shmerl** [comment]
Now that NGG is on in ACO for Navi, should this message still appear?
```
WARNING: disabling NGG because ACO is used.
```

**Comment by Venemo** [comment]
@shmerl As noted [in the merge request](https://gitlab.freedesktop.org/mesa/mesa/merge_requests/3576) and here, we concluded that NGG gives basically zero performance benefits, so we concluded that it's not worth to rush it into the 20.0 release. Instead we'll merge it in 20.1 when not only VS but also GS+VS is ready. There is already a LOT of improvements in ACO since 19.3 so still plenty to get excited about.

**Comment by Venemo** [comment]
Some recent Navi progress: ACO now supports NGG (Next Generation Geometry) for vertex and tessellation evaluation shaders. Our implementation is based on a few ideas from RadeonSI, and is slightly more efficient than what RADV/LLVM does. That being said, I still didn't observe a noticable performance benefit from NGG yet. However, we still need to support it if we want to be future proof.

With this, and our recent addition of tessellation shaders, the only shader stages missing are the merged NGG geometry shaders `ngg_vertex_geometry_gs` and `ngg_tess_eval_geometry_gs`. These are in my plans, but they are not a priority right now.

**Comment by shmerl** [comment]
Any new config keys to enable those?

**Comment by Venemo** [comment]
@shmerl You don't need to do anything to enable them. They will be used automatically, except for NGG GS which is not supported so will fallback to legacy GS.

**Comment by shmerl** [comment]
From the above and some tests I run, there is indeed no noticeable performance impact. Is NGG path itself envisioned by AMD as something faster, or as simply a different hardware path to replace the old one, without the focus on better performance? May be it has some other benefits like better power efficiency?

**Comment by pendingchaos** [comment]
NGG is more flexible and potentially faster if culling is implemented (though afaik this isn't useful for games)

NGG GS can be faster than legacy GS because it eliminates the GSVS ring

**Comment by Venemo** [comment]
@shmerl Speaking of the traditional Vulkan or OpenGL pipeline, NGG is not going to do anything revolutionary. For vertex and tess eval shaders, the same thing happens as before, but the shader program is more explicitly responsible for some hw details (allocating GS space, exporting primitives, etc). For geometry shaders, NGG eliminates a copy to and from VMEM (this is the GSVS ring that Ryhs mentioned), which might give a slight improvement.

However, consider new use cases like mesh shaders and such, along with features like primitive culling, so basically any kind of usage which needs the shader program to have more fine grained control over the vertices and primitives emitted. Those are made very easy to implement with NGG, while the legacy GS stage is not really feasible for these kind of applications.

So, in my opinion, you should think of NGG as an enabler for new and more efficient geometry features, rather than a silver bullet to make your games fast.

**Comment by Venemo** [comment]
I now have [a branch](https://gitlab.freedesktop.org/Venemo/mesa/-/commits/aco-ngg-gs/) which adds support for ACO NGG GS, and works with the sample app and a few games.

Currently missing support for streamout (aka. transform feedback) and shader queries, but those are going to be next.

**Comment by shmerl** [comment]
What is the current state for enabling cswave32 and gewave32. Is it recommended or may be they are on by default?

**Comment by Venemo** [comment]
@shmerl They are still off by default. They may or may not give a performance advantage in some games. We currently don't have a good way to predict whether Wave32 is advantageous for a given shader or not. I still plan to work on this but it's not high priority. (There are other things we can do which give more measurable benefits.)
